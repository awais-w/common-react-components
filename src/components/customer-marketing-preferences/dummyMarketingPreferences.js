export default {
  id: 300,
  email: 'olga@gmail.com',
  consents: [
    {
      mission: 'AFS',
      missionConsents: [
        {
          consent: 'Y',
          communicationChannel: 'Email',
        },
      ],
    },
    {
      mission: 'ARGOS',
      missionConsents: [
        {
          consent: 'N',
          communicationChannel: 'Email',
        },
      ],
    },
  ],
  message:
    "When you opt in, we will send you information on promotions and special offers. This could be through, post, by email, text message or push notifications on our apps. Make sure you don't miss out.",
};
