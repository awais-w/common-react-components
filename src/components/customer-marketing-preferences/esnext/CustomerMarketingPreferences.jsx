import React from 'react';
import PropTypes from 'prop-types';
import bem from '@argos/utils/esnext/bem';
import Toggle from '@cmc/toggle/esnext/Toggle';
import './CustomerMarketingPreferences.scss';
import Button from '@cmc/cha-button/esnext/Button';

const cn = bem({ prefix: 'cmc-', block: 'customer-marketing-preferences' });

const propTypes = {
  /**
   * Email to be updated with preferences
   */
  email: PropTypes.string,
  /**
   * Array of all consent types, each element is it's own mission, AFS, ARGOS, BANK, etc.
   */
  consents: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * Mission types, AFS, ARGOS, BANK, etc.
       */
      mission: PropTypes.string,
      /**
       * Array of mission types, email, phone, etc.
       */
      missionConsents: PropTypes.arrayOf(
        PropTypes.shape({
          /**
           * Communication channel, email, phone, etc.
           */
          communicationChannel: PropTypes.string,
          /**
           * Consent Enum, Y, N, etc.
           */
          consent: PropTypes.string,
        }),
      ),
    }),
  ),
  /**
   * Helper function to update consent data for specific email and missions
   */
  updateConsentData: PropTypes.func.isRequired,
};

const defaultProps = {
  consents: [],
  email: '',
};

const CustomerMarketingPreferences = ({ consents, email, updateConsentData }) => (
  <div className={`${cn()} container`}>
    <p className={cn('message')}>
      When you opt in, we will send you information on promotions and special offers. This could be through, post, by
      email, text message or push notifications on our apps. Make sure you don&apos;t miss out.
    </p>

    <article>
      <ul className={cn('item-container')}>
        {consents.map((consentTypes) =>
          consentTypes.missionConsents.map((missionConsent) => {
            const { consent, communicationChannel } = missionConsent;
            // mission only exists on higher node consentTypes, so is de-structured from consentTypes
            const { mission } = consentTypes;
            const selected = consent === 'Y';
            const toggleLabel = `${mission} - ${communicationChannel}`;

            // Consent is unset, show button
            if (consent === null) {
              const updateConsent = () => {
                // Default initialise to Y
                updateConsentData({
                  email,
                  mission,
                  consent: 'Y',
                });
              };

              return (
                <li key={`${mission}${communicationChannel}-unset`} className={cn('item')}>
                  <span className={cn('item-label')}>{`${toggleLabel} is not set`}</span>
                  <span>
                    <Button onClick={updateConsent} text={'Set Preference'} />
                  </span>
                </li>
              );
            }

            // Consent is set, show toggle
            const updateConsent = (value) => {
              updateConsentData({
                email,
                mission,
                consent: value === true ? 'Y' : 'N',
              });
            };

            return (
              <li key={`${mission}${communicationChannel}`} className={cn('item')}>
                <Toggle selected={selected} onSelect={updateConsent} text={toggleLabel} />
              </li>
            );
          }),
        )}
      </ul>
    </article>
  </div>
);

CustomerMarketingPreferences.propTypes = propTypes;
CustomerMarketingPreferences.defaultProps = defaultProps;
export default CustomerMarketingPreferences;
