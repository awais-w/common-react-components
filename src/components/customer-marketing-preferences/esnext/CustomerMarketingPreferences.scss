@import '@cmc/common-styles/src/scss/_variables.scss';

.cmc-customer-marketing-preferences {
  @include font-standard;
  width: 100%;
  padding: 0;

  &__message {
    @include font-standard;
    margin-bottom: $base;
  }

  &__item-container {
    list-style-type: none;
    padding: 0;
  }

  &__item {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-bottom: $base;
    padding-bottom: $base;
    border-bottom: $small solid $color-grey-dark;
  }

  &__item-label {
    @include font-bold;
  }

  .cmc-CallToActionButtonCP__button {
    padding: $base;
  }
}

@include touch-device-bp1024 {
  .cmc-customer-marketing-preferences {
    &__message {
      @include font-standard;
      margin-bottom: $medium;
    }
    &__item {
      padding-bottom: $medium;
      padding-top: $half;
    }
  }
}
