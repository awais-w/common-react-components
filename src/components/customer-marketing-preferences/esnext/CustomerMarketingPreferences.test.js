import React from 'react';
import { shallow, mount } from 'enzyme';
import CustomerMarketingPreferences from './CustomerMarketingPreferences';
import dummyMarketingPreferences from '../dummyMarketingPreferences';

const defaultProps = {
  email: '',
  consents: [dummyMarketingPreferences.consents[0]],
  updateConsentData: jest.fn(),
};

const setup = ({ props }) => {
  const component = shallow(<CustomerMarketingPreferences {...props} />);

  return {
    wrapper: component,
    getContainer: () => component.find('.cmc-customer-marketing-preferences'),
    getMessageContainer: () => component.find('.cmc-customer-marketing-preferences__message'),
    getMessage: () => component.find('.cmc-customer-marketing-preferences__message').text(),
    getItemContainer: () => component.find('.cmc-customer-marketing-preferences__item-container'),
    getItem: () => component.find('.cmc-customer-marketing-preferences__item'),
  };
};

describe('CustomerMarketingPreferences with default props', () => {
  const component = setup({ props: defaultProps });

  it('should match default snapshot', () => {
    expect(component.wrapper).toMatchSnapshot();
  });

  it('should render a container', () => {
    expect(component.getContainer().exists()).toBe(true);
  });

  it('should render a message container', () => {
    expect(component.getMessageContainer().exists()).toBe(true);
  });

  it('should display the correct marketing preferences message', () => {
    expect(component.getMessage()).toEqual(dummyMarketingPreferences.message);
  });

  it('should render a preferences item container', () => {
    expect(component.getItemContainer().exists()).toBe(true);
  });

  it('should render a single preference', () => {
    expect(component.getItem().length).toBe(1);
  });
});

describe('CustomerMarketingPreferences with varying number of preferences', () => {
  describe('with no preferences', () => {
    const noPreferences = {
      email: dummyMarketingPreferences.email,
      consents: [],
      updateConsentData: jest.fn(),
    };
    const component = setup({ props: noPreferences });

    it('should render no preferences', () => {
      expect(component.getItem().length).toBe(0);
    });
  });

  describe('with two preferences', () => {
    const twoPreferences = {
      email: dummyMarketingPreferences.email,
      consents: dummyMarketingPreferences.consents,
      updateConsentData: jest.fn(),
    };
    const component = setup({ props: twoPreferences });

    it('should render no preferences', () => {
      expect(component.getItem().length).toBe(2);
    });
  });
});

describe('CustomerMarketingPreferences function', () => {
  describe('updateConsent', () => {
    const propsCalledWith = {
      email: defaultProps.email,
      consent: defaultProps.consents[0].missionConsents[0].consent,
      mission: defaultProps.consents[0].mission,
    };

    // Component must be mounted to access function
    const component = mount(<CustomerMarketingPreferences {...defaultProps} />);

    component.find('button[value=true]').simulate('click');

    it('should toggle the marketing preference', () => {
      expect(defaultProps.updateConsentData).toHaveBeenCalled();
    });

    it('should be called with the correct props', () => {
      expect(defaultProps.updateConsentData).toHaveBeenCalledWith(propsCalledWith);
    });

    it('Consent is flipped to Y/N depending on the initial consent value', () => {
      // When consent is set to "Y" originally test we flip it to "N" on our callback
      expect(defaultProps.updateConsentData).toHaveBeenCalledWith(propsCalledWith);

      // When consent is set to "N" originally test we flip it to "Y" on our callback
      const propsWithConsentDefaultedToNo = {
        email: '',
        consents: [dummyMarketingPreferences.consents[1]],
        updateConsentData: jest.fn(),
      };

      const propCallbackStructureWithConsentAsNo = {
        email: defaultProps.email,
        consent: propsWithConsentDefaultedToNo.consents[0].missionConsents[0].consent,
        mission: propsWithConsentDefaultedToNo.consents[0].mission,
      };

      component.setProps(propsWithConsentDefaultedToNo);
      component.find('button[value=false]').simulate('click');

      expect(propsWithConsentDefaultedToNo.updateConsentData).toHaveBeenCalledWith(
        propCallbackStructureWithConsentAsNo,
      );
    });
  });
});
