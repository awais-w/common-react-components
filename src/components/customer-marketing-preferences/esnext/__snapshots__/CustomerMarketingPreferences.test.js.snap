// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerMarketingPreferences with default props should match default snapshot 1`] = `
<div
  className="cmc-customer-marketing-preferences container"
>
  <p
    className="cmc-customer-marketing-preferences__message"
  >
    When you opt in, we will send you information on promotions and special offers. This could be through, post, by email, text message or push notifications on our apps. Make sure you don't miss out.
  </p>
  <article>
    <ul
      className="cmc-customer-marketing-preferences__item-container"
    >
      <li
        className="cmc-customer-marketing-preferences__item"
        key="AFSEmail"
      >
        <Toggle
          no="NO"
          onSelect={[Function]}
          selected={true}
          text="AFS - Email"
          yes="YES"
        />
      </li>
    </ul>
  </article>
</div>
`;
